### DISCUSSION TITLE
Python exercise doubt in Kaggle

### DISCUSSION THREAD

**Discussion by soumyadeepdas28** [dsc-tiu/MLStudyJams2022#4] [Q&A] [answered]
Hi there I 'm Soumyadeep from GDSC-TIU , I could not understand this question Could anyone help me out this question

. 🌶️ (Optional)
In this problem we'll be working with a simplified version of [blackjack](https://en.wikipedia.org/wiki/Blackjack) (aka twenty-one). In this version there is one player (who you'll control) and a dealer. Play proceeds as follows:

The player is dealt two face-up cards. The dealer is dealt one face-up card.
The player may ask to be dealt another card ('hit') as many times as they wish. If the sum of their cards exceeds 21, they lose the round immediately.
The dealer then deals additional cards to himself until either:
the sum of the dealer's cards exceeds 21, in which case the player wins the round
the sum of the dealer's cards is greater than or equal to 17. If the player's total is greater than the dealer's, the player wins. Otherwise, the dealer wins (even in case of a tie).
When calculating the sum of cards, Jack, Queen, and King count for 10. Aces can count as 1 or 11 (when referring to a player's "total" above, we mean the largest total that can be made without exceeding 21. So e.g. A+8 = 19, A+8+8 = 17)

For this problem, you'll write a function representing the player's decision-making strategy in this game. We've provided a very unintelligent implementation below:

def should_hit(dealer_total, player_total, player_low_aces, player_high_aces):

    """Return True if the player should hit (request another card) given the current game
    state, or False if the player should stay.
    When calculating a hand's total value, we count aces as "high" (with value 11) if doing so
    doesn't bring the total above 21, otherwise we count them as low (with value 1). 
    For example, if the player's hand is {A, A, A, 7}, we will count it as 11 + 1 + 1 + 7,
    and therefore set player_total=20, player_low_aces=2, player_high_aces=1.
    """
    
    return False

This very conservative agent always sticks with the hand of two cards that they're dealt.

We'll be simulating games between your player agent and our own dealer agent by calling your function.

Try running the function below to see an example of a simulated game:

**Comment by diffrxction** [comment, marked as answer]
Hey, I really appreciate your participation! 

The first thing to note is how the game starts. You are given two cards as the player and the dealer (opponent) has one card at the very start.

Now, you(player) have the option to either ask for another card ("HIT") as long as you want. Then once you are done with drawing more cards, the dealer starts drawing cards. They also can choose as many as they want and stop. 

The one whose sum of cards exceeds 21 first loses the game. (When either `player_total` or `dealer_total` exceeds 21)

Other conditions:
- Aces or (A) cards can either hold the value 1 or 11 depending on whatever gives you a better chance of winning and keeps your `player_total` under 21.

So the goal here is to write a logical function using simple if else-if and else, and this function will be simulated for over 50000 games and your win percentage will be revealed to you.

So, try to break the problem in different sub-cases and use conditional statements to design such a function, you can easily get an accuracy of around ~`41-42 %.`

If you want to improve the accuracy by a very small margin, you could study blackjack and common strategies when players generally quit the game and you can use those checks in your function.

Try to design and write some logic for this. 

Hope this clears things out a little for you! If you are still stuck then let me know. 😃
